### PR TITLE
Fixes dev agda

### PR DESCRIPTION
**Pull request by JonasHoefer**
Fixes for #140 (pushing directly to `dev-agda` is prohibited)

### CODE REVIEW

File: tool/full-test.sh [hunk 1]
```diff
@@ -429,7 +429,7 @@ step "Building the Agda base library"        \
      "Built Agda base library successfully"  \
      "Failed to build the Agda base library" \
      "Canceled Agda base library build"      \
-     "./tool/check-agda.sh --recompile base/agda"
+     "./tool/check-agda.sh base/agda"
```
Comment (just95) [inline tool/full-test.sh:432]: Why did you remove the `--recompile` flag?

File: src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs [hunk 2]
```diff
@@ -35,14 +39,16 @@ convertTypeDecls
 convertTypeDecls comp = case comp of
   NonRecursive decl   -> convertTypeDecl decl False
   Recursive    [decl] -> convertTypeDecl decl True
-  Recursive _ ->
-    error "Mutual recursive data types are not supported at the moment."
+  Recursive    ds     -> reportFatal $ Message
+    (IR.typeDeclSrcSpan $ head ds)
+    Error
+    "Mutual recursive data types are not supported at the moment."
```
Comment (just95) [inline src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs]:
```suggestion
  Recursive ds ->
    reportFatal
      $  Message (IR.typeDeclSrcSpan $ head ds) Error
      $  "Mutual recursive data types are not supported by the Agda back end "
      ++ "at the moment."
```

File: src/lib/FreeC/Backend/Agda/Converter/FuncDecl.hs [hunk 3]
```diff
@@ -44,8 +48,10 @@ convertFuncDecls (NonRecursive decl  ) = convertFuncDecl decl Nothing
 convertFuncDecls (Recursive    [decl]) = do
   [decArg] <- identifyDecArgs [decl]
   convertFuncDecl decl $ Just decArg
-convertFuncDecls (Recursive _) =
-  error "Mutual recursive functions are not supported at the moment."
+convertFuncDecls (Recursive ds) = reportFatal $ Message
+  (IR.funcDeclSrcSpan $ head ds)
+  Error
+  "Mutual recursive functions are not supported at the moment."
```
Comment (just95) [inline src/lib/FreeC/Backend/Agda/Converter/FuncDecl.hs]:
```suggestion
convertFuncDecls (Recursive ds) = 
  reportFatal
    $  Message (IR.funcDeclSrcSpan $ head ds) Error
    $  "Mutual recursive functions are not supported by the Agda back end "
    ++ "at the moment."
```

File: src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs [hunk 4]
```diff
@@ -35,14 +39,16 @@ convertTypeDecls
...
 
 -- | Converts a data or type synonym declaration.
 --   TODO: Convert type synonyms.
 convertTypeDecl :: IR.TypeDecl -> Bool -> Converter [Agda.Declaration]
-convertTypeDecl (IR.TypeSynDecl _ _ _ _) _ =
-  error "Type synonyms are not supported at the moment."
+convertTypeDecl (IR.TypeSynDecl srcSpan _ _ _) _ = reportFatal
+  $ Message srcSpan Error "Type synonyms are not supported at the moment."
```
Comment (just95) [inline src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs]:
```suggestion
convertTypeDecl (IR.TypeSynDecl srcSpan _ _ _) _ =
  reportFatal
    $  Message srcSpan Error
    $  "Type synonyms are not supported by the Agda back end "
    ++ "at the moment."
```

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 5]
```diff
@@ -13,43 +13,51 @@ module FreeC.LiftedIR.Converter.Type
...
-  -> [LIR.Type]
+  -> Converter [LIR.Type]
 liftFuncArgTypes = maybe liftNonRecFuncArgTypes liftRecFuncArgTypes
 
 -- | Converts the argument types of a non recursive function using @convertType@.
-liftNonRecFuncArgTypes :: [IR.VarPat] -> [LIR.Type]
-liftNonRecFuncArgTypes = map $ fromJust . liftVarPatType
+liftNonRecFuncArgTypes :: [IR.VarPat] -> Converter [LIR.Type]
+liftNonRecFuncArgTypes = mapM $ \pat ->
+  let err = reportFatal $ Message (IR.varPatSrcSpan pat) Error "Expected Type"
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]: Since this error should never occur, I'd make it an `Internal` error.

```suggestion
  let err =
          reportFatal
            $ Message (IR.varPatSrcSpan pat) Internal
            $ "Expected variable pattern to have a type annotation."
```

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 6]
```diff
@@ -61,32 +69,34 @@ liftConArgType ident = markAllDec ident . liftType
...
 liftTypeApp' (IR.TypeApp _ l r       ) ts = liftTypeApp' l (r : ts)
 liftTypeApp' (IR.FuncType srcSpan l r) [] =
-  LIR.FuncType srcSpan (liftType l) (liftType r)
-liftTypeApp' _ (_ : _) =
-  error "liftTypeApp': Only type constructors can be applied!"
+  LIR.FuncType srcSpan <$> liftType l <*> liftType r
+liftTypeApp' _ (_ : _) = reportFatal $ Message
+  NoSrcSpan
+  Error
+  "liftTypeApp': Only type constructors can be applied!"
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]:
```suggestion
liftTypeApp' _ (_ : _) =
  reportFatal
    $ Message NoSrcSpan Internal
    $ "Only type constructors can be applied!"
```

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 7]
```diff
@@ -112,10 +122,11 @@ markAllDec decName (LIR.TypeCon srcSpan name ts dec) = LIR.TypeCon
...
 markOutermostDecreasing (LIR.TypeCon srcSpan name ts _) =
-  LIR.TypeCon srcSpan name ts True
+  return $ LIR.TypeCon srcSpan name ts True
 markOutermostDecreasing (LIR.FreeTypeCon srcSpan t) =
-  LIR.FreeTypeCon srcSpan $ markOutermostDecreasing t
-markOutermostDecreasing _ =
-  error "Outermost argument is not necessarily unique"
+  LIR.FreeTypeCon srcSpan <$> markOutermostDecreasing t
+markOutermostDecreasing _ = reportFatal
+  $ Message NoSrcSpan Error "Outermost argument is not necessarily unique"
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]:
```suggestion
markOutermostDecreasing _ =
  reportFatal
    $  Message NoSrcSpan Error
    $  "Outermost type of decreasing argument is not a "
    ++ "type constructor application."
```

**Review by just95 — CHANGES_REQUESTED** [review]
I've added some suggestions for how to improve the error messages. <https://github.com/FreeProving/free-compiler/pull/140#discussion_r457226408> still needs to be addressed. You could add instructions for how to setup the `agda-lib` file in order to use the base library, for example.

File: tool/full-test.sh (hunk 1 above)
Comment (JonasHoefer) [reply tool/full-test.sh:432]: Compiling the Agda base library takes long time (compared to the other tests in the script). Furthermore Agda notices if files change and the `.agdai` files become invalid. I think this change saves most people working with the compiler some time and should not hurt the test itself, because the CI pipeline always compiles the whole base library.

File: README.md [hunk 8]
```diff
@@ -182,6 +182,12 @@ To compile the base library with that shell script, run the following command in
 > ./tool/compile-coq.sh --recompile base/coq
 > ```
 
+### Agda Base Library
+
+In order to use the Agda code generated by the Free Compiler, the Agda base library and its dependencies have to be installed.
```
Comment (just95) [inline README.md:190]:
```suggestion
In order to use the Agda code generated by the Free Compiler, the Agda base library and its dependencies have to be installed.

```

File: README.md [hunk 9]
```diff
@@ -182,6 +182,12 @@ To compile the base library with that shell script, run the following command in
...
+1. Install the [Agda Standard Library][software/agda-stdlib] according to its installation instructions.
+2. Register the base library by adding the path to [`base/agda/base.agda-lib`][freec/base/agda/base.agda-lib] to `$HOME/.agda/libraries`
```
Comment (just95) [inline README.md]:
```suggestion
2. Register the base library by adding the path to [`base/agda/base.agda-lib`][freec/base/agda/base.agda-lib] to `$HOME/.agda/libraries`.
```

File: src/exe/FreeC/Backend.hs [hunk 10]
```diff
@@ -150,10 +150,33 @@ convertModuleToAgda :: IR.Module -> Application String
...
+ where
+  -- | Creates the string to write to the @.agda-lib@ file.
+  contents :: String -> String
+  contents name =
+    unlines ["name: " ++ name, "include: .", "depend: standard-library base"]
+
+  -- | Path to the @.agda-lib@ file to create and the name of the library.
+  getAgdaLib :: Application (FilePath, String)
+  getAgdaLib = do
+    Just outputDir <- inOpts optOutputDir
```
Comment (just95) [inline src/exe/FreeC/Backend.hs:181]: In order to avoid a runtime error, the generation of the `.agda-lib` file must be disabled if there is no `--output` directory. We should also consider to add a `--no-agda-lib` command line option (or generalize the `--no-coq-project` command line option). Also don't overwrite the file if it exists already.

File: src/lib/FreeC/LiftedIR/Converter/Type.hs [hunk 11]
```diff
@@ -34,7 +34,10 @@ liftFuncArgTypes = maybe liftNonRecFuncArgTypes liftRecFuncArgTypes
 -- | Converts the argument types of a non recursive function using @convertType@.
 liftNonRecFuncArgTypes :: [IR.VarPat] -> Converter [LIR.Type]
 liftNonRecFuncArgTypes = mapM $ \pat ->
-  let err = reportFatal $ Message (IR.varPatSrcSpan pat) Error "Expected Type"
+  let err =
+          reportFatal
+            $ Message (IR.varPatSrcSpan pat) Internal
+            $ "Expected variable pattern to have a type annotation."
   in  liftVarPatType pat >>= maybe err pure
```
Comment (just95) [inline src/lib/FreeC/LiftedIR/Converter/Type.hs]:
```suggestion
  in  liftVarPatType pat >>= maybe err return
```

**Review by just95 — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -355,6 +357,18 @@ The compiler will not overwrite your changes since the `_CoqProject` file is onl
...
+
+When an output directory has been specified using the `--output` option and the output directory does not contain a `.agda-lib` file, it is created automatically.
+The name of the `.agda-lib` file is the same as the output directory.
+The `.agda-lib` file tells Agda that generated code depends on the compilers base library and the [Agda standard library][software/agda-stdlib].
+
+You can safely edit the `.agda-lib` file or supply your own configuration as long as the dependencies remain intact.
+The compiler will not overwrite your changes since the `.agda-lib` file is only written to if it does not exist.
+
+If you don't want the `.agda-lib` file to be generated at all, you can add the `--no-agda-lib` flag.
+However, note that you may not be able to compile the generated Agda code if the `.agda-lib` file is missing.
```
Comment (just95) [inline README.md:371]: Looks good to me. Only the entry in the table of contents is missing.

**Review by just95 — CHANGES_REQUESTED** [review]

**Review by just95 — APPROVED** [review]

**Review by nbun — APPROVED** [review]
